producer.go

```diff
@@ -152,7 +151,51 @@ func (p *segmentProducer) configureSecurity(cfg Config) error {
 	return nil
 }
 
-func (p *segmentProducer) Send(key []byte, payload []byte, id []byte, ackFunc sdk.AckFunc) error {
+func (p *segmentProducer) Send(ctx context.Context, key []byte, payload []byte, id []byte, ackFunc sdk.AckFunc) error {
+	sendErr := p.sendRetryable(ctx, key, payload, id, ackFunc)
+	// If sendErr == nil, the message was successfully added to a batch,
+	// i.e. not actually sent.
+	// Because of that, we invoke ackFunc here only if sendErr != nil.
+	// NB: ackFunc will be invoked in a callback, when the batch is actually sent.
+	if sendErr == nil {
+		return nil
+	}
```

Comment:
Nitpick: we should strive to keep the happy path on the left. I realize the sad path is longer, but nothing out of the ordinary to put into the `if` clause IMO.

Reply:
I can do that, especially if it's a Go code convention. But at least to me, it's more readable this way. The whole point of keeping the happy path to the left (at least from what I know), is the ability to visually focus and skim through code. In this case, the "more complicated" part is kept indented, which, at least to me, is visually "disturbing". Furthermore, when I'm done reading a long if block and read a `return` I have to ask myself "which case does this return cover?"

Reply:
It is somewhat of a convention yes, see [effective go](https://go.dev/doc/effective_go#if):

> The code reads well if the successful flow of control runs down the page, eliminating error cases as they arise.

I'm not saying we need to religiously follow this advice, but be prepared to hear me point it out in a code review if I notice it 😄 There probably are legitimate cases when we can break this convention, but here I'm honestly not sure we should, it doesn't feel like a special case. I personally already got used to reading things like `if err != nil` and mentally ignore the content of the block when I'm only worried about the happy path. If I see `if err == nil` my brain needs a second to figure out what's going on there.

At the end of the day, this is a personal preference so I marked the comment as a nitpick. It's your call, I won't block because of this!

Reply:
Effective Go makes it more than just a personal preference.: ) I'll update the code. Thanks for pointing this out!